src/basic/devnum-util.c

```diff
@@ -57,6 +58,59 @@ int device_path_make_major_minor(mode_t mode, dev_t devnum, char **ret) {
         return 0;
 }
 
+int device_path_make_major_minor_sysfs(mode_t mode, dev_t devnum, char **ret) {
+        _cleanup_free_ char *syspath = NULL, *link = NULL, *fname = NULL;
+        char *devpath = NULL;
+        const char *t;
+        int r;
+
+        /* Generates the /dev/... path given a dev_t. What makes this different
+         * from device_path_make_major_minor is that it works even when udev 
+         * hasn't yet run */
+
+        if (S_ISCHR(mode))
+                t = "char";
+        else if (S_ISBLK(mode))
+                t = "block";
+        else
+                return -ENODEV;
+
+        if (asprintf(&syspath, "/sys/dev/%s/" DEVNUM_FORMAT_STR, t, DEVNUM_FORMAT_VAL(devnum)) < 0)
+                return -ENOMEM;
+
+        r = readlink_malloc(syspath, &link);
+        if (r < 0)
+                return r;
+
+        r = path_extract_filename(link, &fname);
+        if (r < 0)
+                return r;
+
+        devpath = path_join("/dev", fname);
+        if (!devpath)
+                return -ENOMEM;
+
+        struct stat st;
+        if (stat(devpath, &st) < 0) {
+                free(devpath);
+                return -errno;
+        }
+
+        if ((st.st_mode & S_IFMT) != mode) {
+                free(devpath);
+                return -ENODEV;
```

Comment:
i think it would make sense to return ENOTBLK if we expected a block device, but it is something else.

hence:

```c
return S_ISBLK(mode) ? -ENOTBLK : -ENODEV;
```